src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -0,0 +1,223 @@
+/*-------------------------------------------------------------------------
+ *
+ * insert_select_planner.c
+ *
+ * Planning logic for INSERT..SELECT.
+ *
+ * Copyright (c) 2017, Citus Data, Inc.
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "distributed/citus_ruleutils.h"
+#include "distributed/errormessage.h"
+#include "distributed/insert_select_planner.h"
+#include "distributed/multi_executor.h"
+#include "distributed/multi_logical_planner.h"
+#include "distributed/multi_physical_planner.h"
+#include "distributed/multi_router_planner.h"
+#include "distributed/pg_dist_partition.h"
+#include "nodes/makefuncs.h"
+#include "nodes/nodeFuncs.h"
+#include "nodes/parsenodes.h"
+#include "optimizer/planner.h"
+#include "parser/parsetree.h"
+#include "parser/parse_coerce.h"
+#include "parser/parse_relation.h"
+#include "utils/lsyscache.h"
+
+
+static DeferredErrorMessage * DeferErrorIfCoordinatorInsertSelectUnsupported(
+	Query *insertSelectQuery);
+
+
+/*
+ * CreatteCoordinatorInsertSelectPlan creates a query plan for a SELECT into a
+ * distributed table. The query plan can also be executed on a worker in MX.
+ */
+MultiPlan *
+CreateCoordinatorInsertSelectPlan(Query *parse)
+{
+	Query *insertSelectQuery = copyObject(parse);
+
+	RangeTblEntry *subqueryRte = ExtractSelectRangeTableEntry(insertSelectQuery);
+	RangeTblEntry *insertRte = ExtractInsertRangeTableEntry(insertSelectQuery);
+	Oid targetRelationId = insertRte->relid;
+
+	Query *subquery = (Query *) subqueryRte->subquery;
+
+	ListCell *selectTargetCell = NULL;
+	ListCell *insertTargetCell = NULL;
+
+	MultiPlan *multiPlan = CitusMakeNode(MultiPlan);
+	multiPlan->operation = CMD_INSERT;
+
+	multiPlan->planningError =
+		DeferErrorIfCoordinatorInsertSelectUnsupported(insertSelectQuery);
+
+	if (multiPlan->planningError != NULL)
+	{
+		return multiPlan;
+	}
+
+	ReorderInsertSelectTargetLists(insertSelectQuery, insertRte, subqueryRte);
+
+	if (list_length(insertSelectQuery->cteList) > 0)
+	{
+		/*
```

Comment:
This logic seems complicated enough to live in a separate function. Could you move it to another function if it is easy enough to do?